src/nvim/lua/stdlib.c

```diff
@@ -403,6 +410,10 @@ int nlua_setvar(lua_State *lstate)
 
     // Update the value
     tv_copy(&tv, &di->di_tv);
+    if (vimvar != NULL) {
+      set_var_const(vimvar, strlen(vimvar), &tv, false, false);
+      xfree(vimvar);
+    }
```

Comment:
`set_var_const()` does more things, like notify watchers, prevent the copy when vimvar of wrong type etc. I'd put this is an if statement replacing all of 399-422 (new line numbers) when `vimvar` is set. (using `tv_clear()` and early return would be fine)

Reply:
Just noticed that the vvar stuff in `set_var_const` has an early return that always skips the dictwatcher notify logic. Maybe a bug? :thinking:

Reply:
maybe we only need this part

```C
      } else if (v->di_tv.v_type == VAR_NUMBER) {
        v->di_tv.vval.v_number = tv_get_number(tv);
        if (strcmp(varname, "searchforward") == 0) {
          set_search_direction(v->di_tv.vval.v_number ? '/' : '?');
        } else if (strcmp(varname, "hlsearch") == 0) {
          no_hlsearch = !v->di_tv.vval.v_number;
          redraw_all_later(UPD_SOME_VALID);
        }
        return;
```

Reply:
You'll need to whole block to prevent changing their types also.

Reply:
a simple test this also can work

```C
    // Update the value
    tv_copy(&tv, &di->di_tv);

    if (tv.v_type == VAR_NUMBER && striequal(key.data, "hlsearch")) {
      no_hlsearch = !tv.vval.v_number;
    }

```

Reply:
While it works, it duplicates the logic (and misses the call to `redraw_all_later`) and doesn't address being able to change the type (e.g: `vim.v.errors = 3` should not be allowed).

I think ideally it'd be best to factor the `v:` validation logic in `set_var_const` (without calling `set_var_const` itself).

Reply:
hmmm maybe. that's a test I removed `redraw_all_later` .

Reply:
Oh, seems zeer's handling this now at #25325

Reply:
already saw that  so delete branch, If want to solve it, please do it as soon as possible. rather than after others have spent time on it.